src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs

```diff
@@ -243,6 +243,18 @@ private bool DoesExpressionUseDocumentItself(Expression node)
             return false;
         }
 
+        private string PrepareElemMatchFieldName(IFieldExpression fieldExpression)
+        {
+            if (fieldExpression.Document is IFieldExpression documentFieldExpression)
+            {
+                return $"{documentFieldExpression.FieldName}{fieldExpression.FieldName}";
```

Comment:
The lost node is presented in `Document` expression.

Reply:
NOTE: I made a small investigation and it looks like the real problem of this issue is this change:  https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/13/files#r251429792

Without the above change and the change in this ticket, the generated query for this test https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/53/files#diff-28fd5e9de3efd662172dc9b076ed2c10R256 looks like:

    { G : { $elemMatch : { "D" : "Don't", "S.Z" : { $elemMatch : { "H" : 0 } } } } }  // pay attention on "S.Z" which is correct

When the previous fix is applied and without the changes from this ticket (the situation that the user sees now):

   { G : { $elemMatch : { "D" : "Don't", "Z" : { $elemMatch : { "H" : 0 } } } } }   // pay attention on just `Z` which is the wrong query

So, I guess the current fix + previous fix cover all known issues, but I would suggest to create a new ticket to return this line https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/13/files#diff-25128dea722dc78fdf02f511b3ea418dL1658

It's not possible now since some tests which were written in the scope of `CSHARP-1356` doesn't pass without this change. But as far as I remember, my first variant allowed the logic without removing this method: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/13#discussion_r247197507

Reply:
Why is there no `.` between `{documentFieldExpression.FieldName}` and `{fieldExpression.FieldName}`?

Reply:
it's presented in `documentFieldExpression.FieldName`

Reply:
Can you explain what you mean?

Do you mean that the value returned by `documentFieldExpression.FieldName` ENDS with a `.`?

Reply:
yes

Reply:
That is totally surprising to me, but as long as the `.` gets in their one way or another... I guess it's OK.

Reply:
But just to highlight, I see the main problem here is described in this comment: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/53#discussion_r317409755
I don't mind to not worry about it right now, but it should be fixed somehow, or at least we need to investigate this situation more carefully. Before, we removed this method https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/13/files#r251429792 because we didn't find tests which were affected by this method, but this case is a provement that this method was an "valuable" part of this logic

Reply:
Investigation ticket related to this comment: https://jira.mongodb.org/browse/CSHARP-2744